### PR TITLE
Documentation for the first product

### PR DESCRIPTION
**Pull request by diodon**

### CODE REVIEW

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
+# Non-Velocity Aggregated Product
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: How about "Aggregated Time Series Product (non-velocity)"?

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+
+## Objective
+
+This product project will provide:
+
+A comprehensive set of aggregated time-series files for each mooring site and parameter (excluding current velocity), 
+without any interpolation or filtering, except for the exclusion of the out-of-water data. The Quality Control (QC) 
+flags will be preserved. All the (python) code used for the generation of the products openly available on GitHub.
+Clear documentation on the code and the output file format.
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: This file is the documentation, and it's here with the code on GitHub, so the last two sentences are a bit redundant.

Also, here we are documenting the _code_, not the project. So maybe instead of objective, the first thing to describe is the _output_, which is kind of what the first two sentences say (without the "comprehensive set.." bit).

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+Input Files and variables to be aggregated by this project will meet the following requirements:
+
+- File contains data from only one deployment of one instrument;
+- File is a delayed-mode, quality-controlled product (file version label “FV01”);
+- File is compliant with CF-1.6 and IMOS-1.4 conventions;
+- File contains, at the minimum, the variable of interest (VoI), and variables TIME, LATITUDE, LONGITUDE, and either variable NOMINAL_DEPTH or global attribute instrument_nominal_depth;
+- All files to be aggregated are from the same site, and have the same site_code attribute;
+- Variables to be aggregated have TIME as their only dimension (or if LATITUDE and LONGITUDE are included as dimensions, they have size 1);
+- Variables to be aggregated do not represent current velocity (these will be included in another product); and
+- The in-water data are bounded by the global attributes time_deployment_start and time_deployment_end.
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: This is already included under Input below, so I think you can remove this whole section.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+## Input
+
+The aggregation function will accept a list of input files, and the name of the variable of interest (VoI).
+
+The code will be able to aggregate variables and files that meet the following requirements:
+
+- File contains data from only one deployment of one instrument;
+- File is a delayed-mode product, with file version label FV01;
+- File is compliant with CF-1.6 and IMOS-1.4 conventions;
+- File contains, at the minimum, the VoI, and variables TIME, LATITUDE, LONGITUDE, and   NOMINAL_DEPTH;
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: It looks like you copied this from an earlier version of the IP document. The latest version says 
* File contains, at the minimum, the VoI, and variables TIME, LATITUDE, LONGITUDE, and either variable NOMINAL_DEPTH or global attribute instrument_nominal_depth;

I think the rest of the dot points are unchanged.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+- All files to be aggregated are from the same site, and have the same site_code attribute;
+- Variables to be aggregated have TIME as their only dimension (or if LATITUDE and LONGITUDE are included as dimensions, they have size 1);
+- Variables to be aggregated do not represent current velocity;
+- The in-water data are bounded by the global attributes time_deployment_start and time_deployment_end.
+
+This project will not make any corrections to or implement work-arounds for input files that do not meet these requirements. A report on files that cannot be aggregated will be produced. If any files are re-processed by the moorings facility to meet the requirements, these will be incorporated into future versions of the product.
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: Again the rest of the text in this section is talking about the project. I'm not sure that it's directly relevant to the code.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  -path OUTPUT_PATH  path where the result file will be written. Default ./
+
+```
+
+
+
+
+### Input file selection
+
+The input files are accessed using the OPeNDAP protocol via the AODN THREDDS server.  A list of OPeNDAP URLs can be produced by the utility function get_moorings_url(), which queries an index of all published moorings files to obtain the files relevant for a given site and variable of interest (VoI). The list can also be filtered by feature type (e.g time-series or profile) temporal range, file version (FV01), processing mode (real-time/delayed) or a regular expression matched against the file name.
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: This is also outside the scope of what the aggregation code does. When we import the get_moorings_url() code, we'll document it there. So I don't think we need this section either.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+### Input file validation
+
+Before proceeding to the aggregation, each input file will be checked to ensure it meets the requirements (as specified above under Inputs). This will prevent errors during the aggregation process. If any files fail to meet the requirements, a list of all non-compliant files is output, and the code stops. An option can be provided to continue processing only the compliant files.
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: Again, this is out of date. We later changed this to:
> ... Any input files that fail to meet the requirements will be excluded from the aggregation, and their URL listed in a global attribute rejected_files.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+`date_created`: set to the date/time the product file is created;
+
+`history`: set to “<date_created>: Aggregated file created.”;
+
+`keywords`: set to a comma-separated list of the main variable names (“<VoI>, TIME, DEPTH, LATITUDE, LONGITUDE”);
+
+`lineage`: a statement about how the file was created, including a link to the code used, and any input parameters (except the input files, which are listed in the source_file variable);
+
+`title`: "Aggregated time-series product: <VoI> at <sit````e_code> between <time_coverage_start> and <time_coverage_end>"
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: Something's gone weird with the formatting here...

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+The product will be delivered, in netCDF4 format, compliant with the CF-1.6 and IMOS-1.4 conventions, and
+structured according to the [indexed ragged array representation](http://cfconventions.org/cf-conventions/v1.6.0/cf-conventions.html#_indexed_ragged_array_representation).
+
+## Sample Files
+
+Sample files could be downloaded from AODN THREDDS(add link) server.
+
+
+## Distribution
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: Distribution, discovery and access, and maintenance are not relevant to the code, so I would remove them too.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 6 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: I think it is good idea to keep it as someone might what to use the functions by him/her self on a particular set of files.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 9 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: eliminated

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 1 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: OK

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 3 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: removed

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 4 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: fixed

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 5 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: removed

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 7 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: fixed

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 8 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: yeah, fixed...

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 6 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: Sorry were you referring to the Input file selection section? I think we need to remove this section as the function accept a URL list. It doesn't care about how you get the urls

Comment (mhidas) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: Yes, exactly. Remove the "Input file selection" section.

It could be worth mentioning that the code is able to access the input files either locally, or remotely via the OPeNDAP protocol. This could just be added at the end of the "Input" section, above.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+The dimensions of the resulting file  are determined as follows:
+`OBSERVATION`:  the total number of observation records, excluding out-of-the-water data, in all input files;
+`INSTRUMENT`: the number of instruments (i.e. number of files);
+`string256`: a fixed dimension of length 256 for character array variables.
+
+### Variables
+
+The VoI is produced by sequentially concatenating the individual values in each of the input files. The resulting variable has dimension `OBSERVATION`. The VoI’s ancillary_variables, in particular the corresponding quality-control flags, are also included, with dimension OBSERVATION. If the quality control variable is absent in any input file, the corresponding flags in the output file will  be set to 0 (“no QC performed”) and a warning will be raised.
+
+The variable `TIME` from input files is concatenated into a variable TIME(OBS). This could result in a non-uniform time interval.
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: OBS -> OBSERVATION, and again two more times in the next paragraph.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+The `DEPTH` variable from input files is concatenated into a variable `DEPTH(OBS)`. If not present, fill values are stored. `DEPTH_quality_control`, if present, is also included. Where the `DEPTH` variable is absent, the corresponding `DEPTH_quality_control` values are set to 9 (“missing value”).
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: Missing this text from the IP:
> The variables PRES (sea_water_pressure) and PRES_REL (sea_water_pressure_due_to_sea_water) will be aggregated in exactly the same way as DEPTH. Every aggregated file will contain DEPTH, PRES, and PRES_REL, and corresponding quality control flags.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+All output variables with the OBS dimension are sorted in chronological order.
```
Comment (mhidas) [inline aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: This has been changed too:
> All output variables with the INSTRUMENT dimension are sorted in chronological order, and the input files aggregated chronologically, according to the global attribute time_deployment_start.

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 6 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: ok

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 11 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: added

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 10 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: already fixed

File: aodntools/timeseries_products/Documentation/aggregated_timeseries.md (hunk 12 above)
Comment (diodon) [reply aodntools/timeseries_products/Documentation/aggregated_timeseries.md]: missed but now it is in....

**Comment by mhidas** [comment]
Great, thanks!
